src/common/io/src/buffer/buffer_read_datetime_ext.rs

```diff
@@ -61,48 +70,211 @@ where R: BufferRead
+                        let offset = (current_tz_sec - val_tz_sec) * 1000 * 1000;
+                        let mut ts = dt.timestamp_micros();
+                        ts += offset;
+                        // TODO: need support timestamp_micros in chrono-0.4.22/src/offset/mod.rs
+                        // use like tz.timestamp_nanos()
+                        let (mut secs, mut micros) = (ts / 1_000_000, ts % 1_000_000);
+                        if ts < 0 {
+                            secs -= 1;
+                            micros += 1_000_000;
+                        }
+                        Ok(tz.timestamp_opt(secs, (micros as u32) * 1000).unwrap())
+                    };
+                    if self.ignore(|b| b == b'z' || b == b'Z')? {
+                        // ISO 8601 The Z on the end means UTC (that is, an offset-from-UTC of zero hours-minutes-seconds).
+                        if dt.year() < 1000 {
+                            Ok(tz.from_utc_datetime(
+                                &NaiveDate::from_ymd(1000, 1, 1).and_hms(0, 0, 0),
+                            ))
+                        } else {
+                            let current_tz = dt.offset().fix().local_minus_utc();
+                            calc_offset(current_tz.into(), 0, &dt)
+                        }
+                    } else if self.ignore_byte(b'+')? {
+                        self.parse_time_offset(tz, &mut buf, &dt, false, calc_offset)
+                    } else if self.ignore_byte(b'-')? {
+                        self.parse_time_offset(tz, &mut buf, &dt, true, calc_offset)
+                    } else {
+                        less_1000(dt)
+                    }
+                }
+                Err(err) => {
+                    if err.is_invalid_digit() {
+                        // only date part
+                        buf.clear();
+                        less_1000(dt)
+                    } else {
+                        Err(ErrorCode::BadBytes(format!("err with {:?}", err)))
+                    }
+                }
             }
         } else {
-            res
-        };
-        // convert timestamp less than `1000-01-01 00:00:00` to `1000-01-01 00:00:00`
-        if dt.year() < 1000 {
-            Ok(tz.from_utc_datetime(&NaiveDate::from_ymd(1000, 1, 1).and_hms(0, 0, 0)))
+            less_1000(dt)
+        }
+    }
+
+    fn parse_time_offset(
+        &mut self,
+        tz: &Tz,
+        buf: &mut Vec<u8>,
+        dt: &DateTime<Tz>,
+        west_tz: bool,
+        calc_offset: impl Fn(i64, i64, &DateTime<Tz>) -> Result<DateTime<Tz>>,
+    ) -> Result<DateTime<Tz>> {
+        if self.keep_read(buf, |f| (b'0'..=b'9').contains(&f))? != 2 {
+            return Err(ErrorCode::BadBytes(
+                "err with parse timezone, format like this:[+08:00]",
+            ));
+        }
+        let hour_offset: i32 = lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap();
+        if (0..15).contains(&hour_offset) {
```

Comment:
```(0..15).contains(&hour_offset)```  is slower then cmp?

Reply:
`cargo clippy` suggests using contains.

Reply:
> `(0..15).contains(&hour_offset)` is slower then cmp?

you mean this?

```
if (0..15).contains(&hour_offset) { -> if hour_offset >=0 && hour_offset < 15 {
```

Reply:
yes

Reply:
ok, maybe contains have been optimized for this